core/shortcut_registry.ts

```diff
@@ -166,34 +167,46 @@ export class ShortcutRegistry {
    * @param shortcutName The name of the shortcut to remove from the key map.
    */
   removeAllKeyMappings(shortcutName: string) {
-    for (const keyCode in this.keyMap_) {
+    for (const keyCode of this.keyMap.keys()) {
       this.removeKeyMapping(keyCode, shortcutName, true);
     }
   }
 
   /**
    * Sets the key map. Setting the key map will override any default key
    * mappings.
-   * @param keyMap The object with key code to shortcut names.
+   * @param newKeyMap The object with key code to shortcut names.
    */
-  setKeyMap(keyMap: {[key: string]: string[]}) {
-    this.keyMap_ = keyMap;
+  setKeyMap(newKeyMap: {[key: string]: string[]}) {
+    this.keyMap.clear();
+    for (const key in newKeyMap) {
+      this.keyMap.set(key, newKeyMap[key]);
+    }
   }
 
   /**
    * Gets the current key map.
    * @return The object holding key codes to ShortcutRegistry.KeyboardShortcut.
    */
-  getKeyMap(): {[key: string]: KeyboardShortcut[]} {
-    return object.deepMerge(Object.create(null), this.keyMap_);
+  getKeyMap(): {[key: string]: string[]} {
+    const legacyKeyMap: {[key: string]: string[]} = Object.create(null);
+    for (const [key, value] of this.keyMap) {
+      legacyKeyMap[key] = value;
+    }
+    return object.deepMerge(Object.create(null), legacyKeyMap);
   }
 
   /**
    * Gets the registry of keyboard shortcuts.
    * @return The registry of keyboard shortcuts.
    */
   getRegistry(): {[key: string]: KeyboardShortcut} {
-    return object.deepMerge(Object.create(null), this.registry_);
+    const legacyRegistry: {[key: string]: KeyboardShortcut} =
+        Object.create(null);
+    for (const [key, value] of this.registry_) {
+      legacyRegistry[key] = value;
+    }
+    return object.deepMerge(Object.create(null), legacyRegistry);
```

Comment:
Here we do, because there's a test verifying that modifying the returned KeyboardShortcut object doesn't affect the one actually stored in the registry, which it would if we don't used deepMerge to deeply clone it. I don't think it's relevant for getKeyMap() though, so removed it there.

Reply:
but isn't `legacyRegistry` already a clone? because on line 205 you're creating a new object. so i don't see how modifying the return value could affect the value of the underlying `this.shortcuts` map. but i might be misunderstanding the problem

Reply:
It is, but the KeyboardShortcut objects stored in it aren't without the call to deepMerge, so they could still be mutated by the caller and have that affect the contents of the registry.